src/mgm.rs

```diff
@@ -143,47 +147,181 @@ impl MgmAlgorithmId {
     }
 }
 
+/// The algorithm used for the MGM key.
+pub trait MgmKeyAlgorithm:
+    BlockCipherDecrypt + BlockCipherEncrypt + Clone + KeyInit + private::Seal
+{
+    /// The algorithm ID used in APDU packets
+    const ALGORITHM_ID: MgmAlgorithmId;
+
+    /// Implemented by specializations to check if the key is weak.
+    ///
+    /// Returns an error if the key is weak.
+    fn check_weak_key(_key: &Key<Self>) -> Result<()> {
+        Ok(())
+    }
+}
+
 /// Management Key (MGM).
 ///
 /// This key is used to authenticate to the management applet running on
 /// a YubiKey in order to perform administrative functions.
 ///
 /// The only supported algorithm for MGM keys is 3DES.
 #[derive(Clone)]
-pub struct MgmKey([u8; DES_LEN_3DES]);
+pub struct MgmKey(MgmKeyKind);
+
+#[derive(Clone)]
+enum MgmKeyKind {
+    Tdes(MgmKey3Des),
+}
 
 impl MgmKey {
-    /// Generate a random MGM key
-    pub fn generate() -> Self {
-        let mut key_bytes = [0u8; DES_LEN_3DES];
-        OsRng.fill_bytes(&mut key_bytes);
-        Self(key_bytes)
+    /// Generates a random MGM key for the given algorithm.
+    pub fn generate<C: MgmKeyAlgorithm>(rng: &mut impl CryptoRngCore) -> Result<Self> {
+        match C::ALGORITHM_ID {
+            MgmAlgorithmId::ThreeDes => MgmKey3Des::generate(rng).map(MgmKeyKind::Tdes),
+        }
+        .map(Self)
     }
 
-    /// Create an MGM key from byte slice.
+    /// Generates a random MGM key using the preferred algorithm for the given Yubikey's
+    /// firmware version.
+    pub fn generate_for(yubikey: &YubiKey, rng: &mut impl CryptoRngCore) -> Result<Self> {
+        match yubikey.version() {
+            // Initial firmware versions default to 3DES.
+            Version { major: ..=4, .. }
+            | Version {
+                major: 5,
+                minor: ..=6,
+                ..
+            } => MgmKey3Des::generate(rng).map(MgmKeyKind::Tdes),
+            // Firmware 5.7.0 and above default to AES-192.
+            Version {
+                major: 5,
+                minor: 7..,
+                ..
+            }
+            | Version { major: 6.., .. } => Err(Error::NotSupported),
+        }
+        .map(Self)
+    }
+
+    /// Parses an MGM key from the given byte slice.
     ///
     /// Returns an error if the slice is the wrong size or the key is weak.
+    ///
+    /// TODO: Can we distinguish DES from AES-192? Or do we take `C` as a parameter and
+    /// require the caller to know the type of the bytes they are parsing?
```

Comment:
I'm not sure you can take a generic cipher instance since you're trying to map concrete ciphers to an enum. I think you either need some kind of separate algorithm identifier enum you can pass as a parameter, or to have separate constructor functions per algorithm.

Reply:
We can take `C` and match on `C::ALGORITHM_ID`, which is a `MgmAlgorithmId` enum. However, we can't have a "default" generic of "whichever uniquely matches the length" with that approach.

I'm now leaning towards taking an `Option<MgmAlgorithmId>` argument. `None` means "autodetect, returning an error if we can't distinguish by length", and `Some(alg)` means "parse only as `alg`".